core/src/main/scala/org/scalarelational/instruction/Query.scala

```diff
@@ -13,53 +13,40 @@ import scala.language.existentials
+                                                              with JoinSupport[Types, Result] {
   def apply[T, S](column: ColumnLike[T, S]) = ColumnAlias[T, S](column, alias, None, None)
 
-  def fields(expressions: SelectExpression[_]*) = copy[Vector[SelectExpression[_]], QueryResult[_]](expressions = asVector ++ expressions, converter = DSLSupport.DefaultConverter)
-  def fields(expressions: Vector[SelectExpression[_]]) = copy[Vector[SelectExpression[_]], QueryResult[_]](expressions = this.expressions ++ expressions, converter = DSLSupport.DefaultConverter)
-  def withoutField(expression: SelectExpression[_]) = copy[Vector[SelectExpression[_]], QueryResult[_]](expressions = expressions.filterNot(se => se == expression), converter = DSLSupport.DefaultConverter)
-  def clearFields() = copy[Vector[SelectExpression[_]], QueryResult[_]](expressions = Vector.empty, converter = DSLSupport.DefaultConverter)
-
-  def from(table: Table) = copy[Expressions, Result](table = table)
-  def where(condition: Condition) = copy[Expressions, Result](whereCondition = condition)
+  def where(condition: Condition) = copy[Types, Result](whereCondition = condition)
 
-  def join(joinable: Joinable, joinType: JoinType = JoinType.Join) = PartialJoin[Expressions, Result](this, joinable, joinType)
-  def innerJoin(joinable: Joinable) = join(joinable, joinType = JoinType.Inner)
-  def leftJoin(joinable: Joinable) = join(joinable, joinType = JoinType.Left)
-  def leftOuterJoin(joinable: Joinable) = join(joinable, joinType = JoinType.LeftOuter)
+  def limit(value: Int) = copy[Types, Result](resultLimit = value)
+  def offset(value: Int) = copy[Types, Result](resultOffset = value)
 
-  def limit(value: Int) = copy[Expressions, Result](resultLimit = value)
-  def offset(value: Int) = copy[Expressions, Result](resultOffset = value)
-
-  def groupBy(expressions: SelectExpression[_]*) = copy[Expressions, Result](grouping = grouping ::: expressions.toList)
-  def orderBy(entries: OrderBy[_]*) = copy[Expressions, Result](ordering = entries.toList ::: ordering)
+  def groupBy(expressions: SelectExpression[_]*) = copy[Types, Result](grouping = grouping ::: expressions.toList)
+  def orderBy(entries: OrderBy[_]*) = copy[Types, Result](ordering = entries.toList ::: ordering)
 
   def as(alias: String) = copy(alias = Option(alias))
 
-  def convert[NewResult](converter: QueryResult[NewResult] => NewResult) = copy[Expressions, NewResult](converter = converter)
-  def map[NewResult](converter: Result => NewResult) = {
-    copy[Expressions, NewResult](converter = (qr: QueryResult[NewResult]) => converter(this.converter(qr.asInstanceOf[QueryResult[Result]])))
-  }
+  def map[NewResult](converter: Result => NewResult) = copy[Types, NewResult](converter = this.converter.andThen(converter))
+  def convert[NewResult](converter: QueryResult => NewResult) = copy[Types, NewResult](converter = converter)
 
   def result = new QueryResultsIterator(table.datastore.exec(this), this)
```

Comment:
I think we should make it explicit that `result` has a side-effect. Perhaps `exec` would be a better name?